src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
+            public IsPatternTranslator(LocalRewriter localRewriter, BoundExpression loweredInput)
+            {
+                this._localRewriter = localRewriter;
+                this._factory = localRewriter._factory;
+                this._tempAllocator = new DagTempAllocator(localRewriter._factory);
+                this._loweredInput = loweredInput;
+                this._inputTemp = new BoundDagTemp(loweredInput.Syntax, loweredInput.Type, null, 0);
+                this._conjunctBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+                this._sideEffectBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+            }
 
-                default:
-                    throw ExceptionUtilities.UnexpectedValue(loweredPattern.Kind);
+            public void Dispose()
+            {
+                _conjunctBuilder.Free();
+                _sideEffectBuilder.Free();
+                _tempAllocator.Dispose();
             }
-        }
 
-        BoundPattern LowerPattern(BoundPattern pattern)
-        {
-            switch (pattern.Kind)
+            public class DagTempAllocator : IDisposable
             {
-                case BoundKind.DeclarationPattern:
-                    {
-                        var declPattern = (BoundDeclarationPattern)pattern;
-                        return declPattern.Update(declPattern.Variable, VisitExpression(declPattern.VariableAccess), declPattern.DeclaredType, declPattern.IsVar);
-                    }
-                case BoundKind.ConstantPattern:
+                private readonly SyntheticBoundNodeFactory _factory;
+                private readonly PooledDictionary<BoundDagTemp, BoundExpression> _map = PooledDictionary<BoundDagTemp, BoundExpression>.GetInstance();
+                private readonly ArrayBuilder<LocalSymbol> _temps = ArrayBuilder<LocalSymbol>.GetInstance();
+
+                public DagTempAllocator(SyntheticBoundNodeFactory factory)
+                {
+                    this._factory = factory;
+                }
+
+                public void Dispose()
+                {
+                    _temps.Free();
+                    _map.Free();
+                }
+
+                public BoundExpression GetTemp(BoundDagTemp dagTemp)
+                {
+                    if (!_map.TryGetValue(dagTemp, out var result))
                     {
-                        var constantPattern = (BoundConstantPattern)pattern;
-                        return constantPattern.Update(VisitExpression(constantPattern.Value), constantPattern.ConstantValue);
+                        // PROTOTYPE(patterns2): Not sure what temp kind should be used for `is pattern`.
+                        var temp = _factory.SynthesizedLocal(dagTemp.Type, syntax: _factory.Syntax, kind: SynthesizedLocalKind.SwitchCasePatternMatching);
+                        _map.Add(dagTemp, _factory.Local(temp));
+                        _temps.Add(temp);
+                        result = _factory.Local(temp);
```

Comment:
>result = _factory.Local(temp) [](start = 24, length = 29)

It looks like we are duplicating the node. Consider moving this line before the ```_map.Add``` and using ```result``` in the ```_map.Add``` call.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156539704](https://github.com/dotnet/roslyn/pull/23209#discussion_r156539704) [](ancestors = 156539704)